Start assembling Alonzo transactions and witnesses.

nc6 opened the pull request:
Early work setting up Alonzo types:
- transaction inputs and outputs
- witnesses
- scripts and data

nc6 committed:
Early work setting up Alonzo types:

- transaction inputs and outputs
- witnesses
- scripts and data

polinavino commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Data.hs:
```diff
@@ -0,0 +1,46 @@
+import Shelley.Spec.Ledger.Hashing (HashAnnotated (..))
+
+-- | TODO this should be isomorphic to the plutus type
+data Data era
```
Should it be imported from the Plutus library? Did we ever conclude the discussion on how close this is going to be to plain CBOR?

polinavino replied:
I think Plutus will have or already has a particular single module that (re?)-exports everything that should be accessible to the ledger (this type, the interpreter calling function, etc.). 

I guess things there will also have to become era-parametrized? Or do we maybe want to make this a type family so that only the family is parametrized by era, but not the specific types to which the family is instantiated for each era

polinavino commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs:
```diff
@@ -0,0 +1,68 @@
+    exUnitsSteps :: !Word64
+  }
+  deriving (Eq, Show, Ord)
+
```
we need to have a discussion about how we actually handle this execution units tracking. What is in the spec now (ie this structure) is kind of middle ground between two approaches not finalized and needs to change. @WhatisRT we should schedule a meeting for some time in the next week or two

polinavino commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxWitness.hs:
```diff
@@ -0,0 +1,289 @@
+
+data ScriptDataRaw era = ScriptDataRaw
+  { _scriptDataScripts :: Map (ScriptHash era) (Core.Script era),
+    _scriptDataData :: Set (Data era),
```
Using the correct plural of Datum here seems slightly more confusing than using the wrong one to make it clear that this is a collection of Datum objects (Datums) heeehee

polinavino replied:
also, should we make the scriptDataData and scriptDataScripts consistent with each other - as in, either have them both be a map from hash to the preimage, or have them both be just the preimage? And if so, which way do you prefer?

nc6 replied:
I'm fine with either. The spec currently has them just being the preimage, but the existing Shelley code has them as the map. As yet I haven't grokked the usage patterns well enough to proffer an opinion on what would be best for `Data`. If we want them consistent, then I can happily change this to a Map.

nc6 committed:
Remove superfluous `Main` generated by cabal.

nc6 committed:
Alonzo testing

Start establishing the framework for Alonzo tests, so far just including
roundtrip tests. We include a roundtrip CBOR test for TxWitness.

nc6 committed:
Use internal package ref to test lib.

This is needed to work around a bug with cabal-to-nix. See this thread
(https://input-output-rnd.slack.com/archives/CG386Q1NE/p1607468183215300)
for further details.

TimSheard commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs:
```diff
@@ -0,0 +1,78 @@
+import Data.Word (Word64)
+import GHC.Generics (Generic)
+import NoThunks.Class (NoThunks)
+
+-- | Marker indicating the part of a transaction for which this script is acting
+-- as a validator.
+data Tag
+  = -- | Validates spending a script-locked UTxO
+    Input
+  | -- | Validates minting new tokens
+    Mint
+  | -- | Validates certificate transactions
+    Cert
+  | -- | Validates withdrawl from a reward account
+    Wdrl
+  deriving (Eq, Generic, Ord, Show)
```
Lets put a comment here about how we intend to use this. Is it as a index to some type constructor (using Data.Kinds) or as a field to some record type.

TimSheard commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxWitness.hs:
```diff
@@ -0,0 +1,330 @@
+    witsData,
+    witsRdmr
+  } <-
+  TxWitnessConstr
+    ( Memo
+        ( TxWitnessRaw
+            witsVKey
+            witsBoot
+            ( ScriptDataConstr
+                ( Memo
+                    ( ScriptDataRaw
+                        witsScript
+                        witsData
+                        witsRdmr
+                      )
+                    _
+                  )
+              )
+          )
+        _
+      )
```
Shouldn't we use the constructor patterns for ScriptData, rather than building it from whole cloth here?

TimSheard reviewed (commented):
Things look good, but I have some questions about serialization of the Scripts and Witnesses.
perhaps you could talk me through it.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs:
```diff
@@ -0,0 +1,438 @@
+  ) =>
+  FromCBOR (TxBodyRaw era)
+  where
+  fromCBOR = decode $ SparseKeyed "TxBodyRaw" init bodyFields []
```
The final argument to `SparseKeyed` here is the set of required fields, such that an error will be raised if they're not present. So should set it here to the same set as in shelley-ma

TimSheard reviewed (approved):
After our discussion clarifying how TxWitness and TxScript are used this looks good.

JaredCorduan reviewed (approved):
I didn't realize I hadn't approved this already, sorry. Thanks @nc6 !
